### PR TITLE
Add Support Layer List to ASP

### PR DESCRIPTION
**Pull request by mingxu1067**
### PR types

New features
### PR changes

APIs
### Describe

1. Added a function to allow user to add their own layer and correspond pruning way to ASP support.
2. Changed naming rule of asp masks to make it clear.
3. Usage example
```python
import numpy as np
import paddle
from paddle.static import sparsity

paddle.enable_static()

def special_pruning(tensor, m, n, mask_algo, param_name):
    static_tensor = np.random.rand(*tensor.shape).astype(np.float32)
    static_tensor_mask = np.random.rand(*tensor.shape).astype(np.float32)
    return static_tensor, static_tensor_mask

main_program = paddle.static.Program()
startup_program = paddle.static.Program()

with paddle.static.program_guard(main_program, startup_program):
    input_data = paddle.static.data(name='data', shape=[None, 128])
    label = paddle.static.data(name='label', shape=[None, 10])
    hidden = paddle.static.nn.fc(x=input_data, num_flatten_dims=-1, 
                                 size=32, activation=None, name="need_special_prune")
    hidden = paddle.static.nn.fc(x=hidden, num_flatten_dims=-1, size=32, activation=None)
    prob = paddle.static.nn.fc(x=hidden, num_flatten_dims=-1, size=10, activation=None)
    loss = paddle.mean(paddle.nn.functional.square_error_cost(prob, label))


sparsity.add_supported_layer("need_special_prune", special_pruning)

place = paddle.CPUPlace()
exe = paddle.static.Executor(place)
exe.run(startup_program)

sparsity.prune_model(main_program, mask_algo="mask_1d", with_mask=False)
```

### CODE REVIEW

File: python/paddle/fluid/contrib/sparsity/__init__.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
-# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
-# Copyright (c) 2021 NVIDIA Corporation.  All rights reserved.
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
+# Copyright (c) 2022 NVIDIA Corporation.  All rights reserved.
```
Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/__init__.py]: 该文件不是本PR新增，不用修改CopyRight年份，下同。这是误改？

File: python/paddle/fluid/contrib/sparsity/__init__.py [hunk 2]
```diff
@@ -29,10 +29,11 @@
...
 from .asp import set_excluded_layers
 from .asp import reset_excluded_layers
+from .supported_layer_list import add_supported_layer
 
 __all__ = [
     'calculate_density', 'check_mask_1d', 'get_mask_1d', 'check_mask_2d',
     'get_mask_2d_greedy', 'get_mask_2d_best', 'create_mask', 'check_sparsity',
     'MaskAlgo', 'CheckMethod', 'decorate', 'prune_model', 'set_excluded_layers',
-    'reset_excluded_layers'
+    'reset_excluded_layers', 'add_supported_layer'
```
Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/__init__.py:38]: `add_supported_layer`、`set_excluded_layers`、`reset_excluded_layers`，这三个接口有点令人迷惑了。

Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/__init__.py:38]: `add_supported_layer`、`set_excluded_layers`、`reset_excluded_layers`，这三个接口有点令人迷惑了。

File: python/paddle/fluid/contrib/sparsity/asp.py [hunk 3]
```diff
@@ -293,7 +295,7 @@ class ASPHelper(object):
     """
 
     MASK_APPENDDED_NAME = '_asp_mask'
-    SUPPORTED_LAYERS = {'fc': 'w_0', 'linear': 'w_0', 'conv2d': 'w_0'}
+    PADDLE_WEIGHT_SUFFIX = "w_"
```
Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/asp.py:298]: 这样改的好处，是可以支持更多的layer，不限定于`fc`、`linear`、`conv2d`？

Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/asp.py:298]: 这样改的好处，是可以支持更多的layer，不限定于`fc`、`linear`、`conv2d`？

File: python/paddle/fluid/contrib/sparsity/asp.py [hunk 4]
```diff
@@ -384,7 +379,7 @@ def _get_mask_name(param_name):
         Returns:
             string: The mask name of :attr:`param_name`.
         """
-        return param_name + ASPHelper.MASK_APPENDDED_NAME
+        return param_name + "." + ASPHelper.MASK_APPENDDED_NAME
```
Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/asp.py:382]: 这里加了`.`，还需要保留`_`吗？

Comment (Xreki) [inline python/paddle/fluid/contrib/sparsity/asp.py:382]: 这里加了`.`，还需要保留`_`吗？

File: python/paddle/static/sparsity/__init__.py [hunk 5]
```diff
@@ -18,11 +18,13 @@
...
+from ...fluid.contrib.sparsity import add_supported_layer  #noqa: F401
 
 __all__ = [     #noqa
     'calculate_density',
     'decorate',
     'prune_model',
     'set_excluded_layers',
-    'reset_excluded_layers'
+    'reset_excluded_layers',
+    'add_supported_layer'
```
Comment (Xreki) [inline python/paddle/static/sparsity/__init__.py]: 这里后续PR再加？`add_supported_layer`、`set_excluded_layers`、`reset_excluded_layers`，这三个接口有点令人迷惑了。

Comment (Xreki) [inline python/paddle/static/sparsity/__init__.py]: 这里后续PR再加？`add_supported_layer`、`set_excluded_layers`、`reset_excluded_layers`，这三个接口有点令人迷惑了。

File: python/paddle/fluid/contrib/sparsity/__init__.py (hunk 1 above)
Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/__init__.py]: Done, 2022 -> 2021

File: python/paddle/fluid/contrib/sparsity/__init__.py (hunk 2 above)
Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/__init__.py:38]: 這三個接口功能分別如下
add_supported_layer: 讓使用者增加自己客製化的nn.Layer的參數到ASP的支援清單，預設只會支援FC, Conv2D, Linear.
set_excluded_layers: 在支援的nn.Layer的參數中，針對特定的幾層不執行ASP, 比如 一個網絡有三層FC, 但其中一層不需要ASP，怎可以使用此API來將該層排除
reset_excluded_layers: 將 `set_excluded_layers` 所添加的Layer全部清空。

Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/__init__.py:38]: 同上

File: python/paddle/fluid/contrib/sparsity/asp.py (hunk 3 above)
Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/asp.py:298]: 是的，此次改動將支援的Layer改為統一由supported_layer_list來管理

Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/asp.py:298]: 同上

File: python/paddle/fluid/contrib/sparsity/asp.py (hunk 4 above)
Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/asp.py:382]: Done, remoe `_`.

Comment (mingxu1067) [reply python/paddle/fluid/contrib/sparsity/asp.py:382]: 同上

File: python/paddle/static/sparsity/__init__.py (hunk 5 above)
Comment (mingxu1067) [reply python/paddle/static/sparsity/__init__.py]: 同上解釋後，是否還是要暫緩對外公開呢？

Comment (mingxu1067) [reply python/paddle/static/sparsity/__init__.py]: 同上解釋後，是否還是要暫緩對外公開呢？

Comment (mingxu1067) [reply python/paddle/static/sparsity/__init__.py]: 如線下討論，暫緩公開

**Review by Xreki — APPROVED** [review]
LGTM

**Review by lanxianghit — APPROVED** [review]

**Review by TCChenlong — APPROVED** [review]
